### PR TITLE
Location Based Restrictions

### PR DESCRIPTION
**Pull request by jakobrehmann**
This PR adds the functionality of location based restrictions. The allows different districts or neighborhoods to have varying values for activity reduction. For instance, the activity reduction for workplaces in Berlin-Spandau may be different than for workplaces in Berlin-Mitte. So far, these location based restrictions kick in for activities, but not yet for the trips between activities.

### CODE REVIEW

File: src/main/java/org/matsim/episim/model/AbstractContactModel.java [hunk 1]
```diff
@@ -73,15 +75,31 @@
 	 */
 	private double curfewCompliance;
 
+	private Map<String, String> subdistrictFacilities;
```
Comment (rakow) [inline src/main/java/org/matsim/episim/model/AbstractContactModel.java:78]: Could be final maybe?

File: src/main/java/org/matsim/run/modules/SnzBerlinScenario25pct2020JR.java [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Snz scenario for Berlin.
+ *
+ * @see AbstractSnzScenario
+ */
+public final class SnzBerlinScenario25pct2020JR extends AbstractSnzScenario2020 {
```
Comment (rakow) [inline src/main/java/org/matsim/run/modules/SnzBerlinScenario25pct2020JR.java:52]: Do you need a copy of the whole file? It is possible to extend/delegate to the existing scenario and only make additional changes?

File: src/main/java/org/matsim/scenarioCreation/ZipcodeAndSubdistrictLookup.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+
+
+		return 0;
+	}
+
+	/**
+	 * Helper class to provide an index for a shapefile lookup.
+	 */
+	public static final class Index {
```
Comment (rakow) [inline src/main/java/org/matsim/scenarioCreation/ZipcodeAndSubdistrictLookup.java:218]: This class already exists somewhere else. Please try to reuse the existing one.

**Review by rakow — CHANGES_REQUESTED** [review]
Please see individual comments for some remarks. Also the tests need to pass first.

**Review by rakow — COMMENTED** [review]
Looks good so far! Could you add/extend the Restriction test to account for the new
districtSpecificValue?

**Comment by jakobrehmann** [comment]
Thanks for your feedback and review @rakow . locationBasedRestriction functionality is now tested on various levels:

1) Restriction: checks that cloning, updating, and merging of works correctly for district specific restrictions;

2) FixedPolicy: checks that the config builder correctly builds fixed policies; and

3) DefaultContactModel: Checks that the districtSpecificValue is used for remaining fraction if a container is in the corresponding district, and that the default remaining fraction is used if not.

The location based functionality still lives within the AbstractContactModel, so it should be theoretically available to all contact models; however, the constructor for each implementation must be changed to accept a Scenario (containing activity facilities) for the functionality to perform as intended. I've done this change for the DefaultContactModel and the SymmetricContactModel. Is there a more elegant solution you can think of?

**Comment by rakow** [comment]
Is it ready to merge? Still marked as draft currently.

**Comment by jakobrehmann** [comment]
Several things have been updated:
- location based restrictions can now be processed when ActivityHandling==startOfDay. This was done by adding a LocationBasedParticipationModel. A corresponding test was also created. 
- The merging behavior for locationBasedRf was modified: If user manually adds restriction for remaining fraction, the previous locationBasedRf will be deleted. For instance, if work activity levels are manually reduced for a holiday, the old locationBasedRf should not kept, since it still contains the higher activity levels.  
- In CreateRestrictionsFromCSV, the extrapolation function was expanded with respect to location based restrictions.
- districtSpecificValue was refactored to locationBasedRf for clarity-sake.

**Review by jakobrehmann — COMMENTED** [review]
it can be removed

**Review by rakow — APPROVED** [review]
Very good work, thank you @jakobrehmann !

**Pull request merged**